### DISCUSSION TITLE
New! Full Support for Google Vertex AI Gemini Pro and Pro Vision

### DISCUSSION THREAD

**Discussion by tzolov** [spring-projects/spring-ai#371] [Announcements]
Find more about Spring AI - Vertex AI Gemini: 
* [VertexAI Gemini Chat](https://docs.spring.io/spring-ai/reference/0.8-SNAPSHOT/api/clients/vertexai-gemini-chat.html)
* [Gemini Function Calling](https://docs.spring.io/spring-ai/reference/0.8-SNAPSHOT/api/clients/functions/vertexai-gemini-chat-functions.html)
* [Support of Multimodal chat requests](https://github.com/spring-projects/spring-ai/blob/65d42c9d4fb4a434a6c262747a86a2af1e82e05e/models/spring-ai-vertex-ai-gemini/src/test/java/org/springframework/ai/vertexai/gemini/VertexAiGeminiChatClientIT.java#L188)
* [Native AOT support](https://github.com/tzolov/spring-ai-aot-tests/tree/main/vertexai-gemini-aot-demo)

**Comment by DenisStrazhevich** [comment]
Great job! @tzolov Can you please tell me if there are plans to integrate Google AI to work with models (e.g. Gemini) without a Google Cloud account, but with an API key?

**Comment by tzolov** [comment]
This is a great question @DenisStrazhevich and perhaps you can help me with it a bit. 
I've added already [support](https://github.com/spring-projects/spring-ai/blob/8701cccfabfd72abee4a9da31e5bf5f74c07a7c9/spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/vertexai/gemini/VertexAiGeminiConnectionProperties.java#L46) passing the credential JSON file as a `spring.ai.vertex.ai.gemini.credentialsUri`, that when provided are used to create 

```java
GoogleCredentials credentials = GoogleCredentials.fromStream(connectionProperties.getCredentialsUri().getInputStream());
return new VertexAI(connectionProperties.getProjectId(), connectionProperties.getLocation(), credentials);
``` 
In the [VertexAiGeminiAutoConfiguration#vertexAi](https://github.com/spring-projects/spring-ai/blob/8701cccfabfd72abee4a9da31e5bf5f74c07a7c9/spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/vertexai/gemini/VertexAiGeminiAutoConfiguration.java#L49C18-L49C26).

However I wan't able to property verify this as I failed to extract the right credentials for my gemini account. 

May you know (or have understandable docs) how to generate the credentials files and eventually the api-key from an existing service? 
I'm a bit lost in google's console and studio :/ But if i figure out how to obtain those for my test account it will be trivial to implement the authentication.

**Reply by DenisStrazhevich** [reply]
@tzolov I think [it should help](https://developers.google.com/workspace/guides/create-credentials). It describes how to generate a credentials.json file in your Google Cloud console.

But my question was a little bit about something else, I put it wrong.

At the moment, you have implemented support for the [Vertex AI Gemini API](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/gemini#gemini-pro) only for the Google Cloud Platform users.
This API requires specifying a projectId and location for authentication purposes, aligning with the cloud-based infrastructure.

However, I've found an alternative approach to interact with Gemini using [Google AI Gemini API](https://ai.google.dev/tutorials/rest_quickstart), that does not necessitate cloud dependency. This method involves generating an API key directly from Google AI Studio, which can then be seamlessly incorporated into requests either as a header or as a query parameter. 

Should this be implemented as part of a separate starter, following the example of OpenAI and Azure OpenAI?

**Reply by prakash-aathi** [reply]
@DenisStrazhevich did you find any  alternative  solution ? is that spring Ai project supports gemini API key?
